include/storage/fbw.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <drivers/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*fbw_callback_t)(u8_t *buf, size_t len, size_t offset);
+
+/**
+ * @brief Structure for flash buffered writes
+ *
+ * Users should treat these structures as opaque values and only interact
+ * with them through the below API.
+ */
+struct fbw_ctx {
+	u8_t *buf; /* Write buffer */
+	size_t buf_len; /* Length of write buffer */
+	size_t buf_bytes; /* Number of bytes currently stored in write buf */
+	struct device *fdev; /* Flash device */
+	size_t bytes_written; /* Number of bytes written to flash */
+	size_t offset; /* Offset from base of flash device to write area */
+	size_t available; /* Available bytes in write area */
+	fbw_callback_t callback; /* Callback invoked on completed write op */
+#ifdef CONFIG_FBW_ERASE
+	off_t last_erased_page_start_offset; /* Last erased offset */
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * api-tags: pre-kernel-ok
+ *
+ * @param ctx context to be initialized
+ * @param fdev Flash device to operate on
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device to start writing to
+ * @param size Number of bytes available for performing buffered write.
+ *             If this is '0', the size will be set to the total size
+ *             of the flash device minus the offset.
+ * @param cb Callback to be invoked on completed flash write operations.
+ *           This can be NULL, in which case no callback is invoked.
+ *           If a callback is set, it is invoked every time a flash_write()
+ *           operation has completed. flash_read() is used to read the data
+ *           written in the last flash_write(). The data is stored in the same
+ *           buffer used for writing data ('buf'). A pointer to this buffer,
+ *           the length of the written data, and the offset written to is sent
+ *           as an argument to the callback.
+ *
+ * @return  0 on success, negative errno code on fail
```

Comment:
positive is not defined! Same below on few places.

Reply:
I'm not sure I understand what you mean, should I explicitly state that positive return values are not used?

Reply:
maybe "0 on success, non-zero on failure"?

Reply:
IMO it is better to state that the return value is negative. This way a caller which forwards the error code knows if it should be negated or not, to be made negative.

Also this seems to be the idiomatic way to describe this type of return code, it is used around 330 times AFAICS.

Reply:
I'm pushing for "non-negative for success, negative for error" to be used generally.  This provides a binary decision to the caller (no question about success/failure) while allowing for distinguishing types of success, even if the specific values aren't part of the public API.  It's a nice way to get internal state information out of functions in unit tests.

Reply:
If we go for this we can no longer use `if (ret)` which is a common way of checking if the return value indicates success or not

```
➜  zephyr git:(fbw) ✗ rg -t c "if \((ret|rc|retval|ret_bytes)\)" | wc   
    655    2631   29820
➜  zephyr git:(fbw) ✗ rg -t c "if \((ret|rc|retval|ret_bytes)[^\)]" | wc 
   1903   11740  102000
```

Reply:
Updated to your proposal @pabigot

Reply:
It's still common in existing code, but it's also in violation of MISRA (treating an `int` value as a boolean), which is why you may see new code use `if (ret != 0)`.  Then `if (ret < 0)` becomes an easy difference.

Reply:
Good point. I changed all tests from flash driver to `if (rc != 0)` as the API does not define whether positive value is success or not.

When testing the `rc` of `fbw_erase` I use `if (rc < 0)` as positive is documented as success in this case.